### ISSUE TITLE
[Feature request] Filter feeds by total count per day/week/month

### ISSUE THREAD

**Issue by f6ADt** [newsboat/newsboat#1367]
Illustration example:
Feeds of larger platforms with several articles per day or week are combined together with blog feeds, whose authors post by far more rarely, like once or twice per month.
Now, when I create a custom query feed of unread articles for a certain tag, there is a risk, these blog feeds get hidden due to the sheer amount of other articles in larger feeds.

I would like to filter/partition (unread) articles based on published articles per day/week/month of a feed. Being quite new to Newsboat, I encountered [`total_count`](https://newsboat.org/releases/2.21/docs/newsboat.html#attr-total_count), but this attribute seems to take into account only the total count of  all feeds. I guess, creating these number from the cache could be done quite easily.

Before in Feedly, I created separate categories for a topic, like "Code/100_per_week", "Code/10_per_week", "Code/1_per_month". So I needed to lookup manually, how active the feed is - fortunately Feedly can spit this number out comfortably.

How can I (how do others) do this in Newsboat? Is this use case compelling enough to implement/extend the `total_count` attribute with some time-based dimension - or should I rather manually create tags for "100/week" etc. again?

**Comment by Minoru** [comment]
Interesting, thank you for the suggestion! If such a feature was implemented, there are a couple implications that I can think of:
- we should also provide a way to display the number for an individual feed, so the user can figure out which "buckets" they want to create
- we need to add real number support to the filter language, since the user would have to write "0.3/week" to filter for monthly feeds
- it's an open question how to calculate the "posting rate".

    A simple average will not work if the feed changed its schedule anywhere in between (or appears to have done so because of `ignore-article`).

    A simple average also raises a question of the period over which it should be calculated:

    1. from first article to last? Then dead feeds would be forever pinned at a particular frequency.
    2. from first article to today? Now the frequency of a dead feed will gradually decrease, but it won't ever reach zero.
    3. over the last 365 days? Now dead feeds will reach zero frequency in a year — just like feeds that *are* active but post once a few years (as a feed for e.g. [HOPL](https://en.wikipedia.org/wiki/History_of_Programming_Languages) might).

    There are other varieties of averages, too. Exponential decay might be a good candidate here, but the "365 days" problem will be back, just phrased as a decay coefficient this time.

Also: I don't think Newsboat should dictate users how to consume their feeds, but I wonder how well this categorization approach works. It seems to use posting frequency as a signal of interestingness/usefulness/relevancy, which in my experience doesn't hold: I have a daily updated feed from a site like bash.org which I do find interesting and read every day, but I also have news feeds from a few webhosting providers which are updated rarely and I only skim, and also high-freq/low-interest news feeds and low-freq/high-interest blogs. @f6ADt, do you have any "outliers" like this? Are you really driver by those numbers, or do you use them as a hint of how valuable the feed is?

(I'm on the fence about this, and think more discussion is needed, so labeling as "to consider" for now.

**Comment by Minoru** [comment]
Oh, one more thing:

> I encountered total_count, but this attribute seems to take into account only the total count of all feeds

Looking at the code, `total_count` is only handled in a single place, and there it returns the number of items in one particular feed: https://github.com/newsboat/newsboat/blob/119582d72d2dd25993d06b6f809e26944827237e/src/rssfeed.cpp#L153-L154 If your experiments show that it returns the count for *all* feeds, please file an issue with steps to reproduce this and we'll see what's going on.

**Comment by f6ADt** [comment]
> we should also provide a way to display the number for an individual feed, so the user can figure out which "buckets" they want to create

Yes, ideally there is one measure/number for each feed that states, how "active" this feed is, so you can use it as filter criteria.

> we need to add real number support to the filter language, since the user would have to write "0.3/week" to filter for monthly feeds

Yeah, you might set 
a) time interval like `per/week` 
b) a number argument `N` to only include feeds with a *maximum* of `N` articles per time interval. 

E.g. include all feeds with max. N=3 articles per week. Good idea to take week as the only interval and leave out/derive day/month.

> it's an open question how to calculate the "posting rate".

In a first step, this could be kept as simple as possible: For each feed, count *all* stored/fetches articles grouped per week, and calculate the average (later on other options possible) and cache this measure as data basis for query filters.

To mitigate recent changes of the feed's activity/schedule, the user might either delete older articles (e.g. `max_items`; IMO perfectly fine for the start). Or add an additional date argument stating, that a feed's activity calculation should start at this particular date. In latter case, you would need to rethink caching mechanisms a bit.

Concerning dead feeds: These are certainly hard to distinguish from "rarely posted" feeds. Though IMO this is more of an edge case and could be solved manually. If there is a dead feed, drop it directly from the list. Or give rare feeds a unique tag to distinguish them from the dead ones.

> It seems to use posting frequency as a signal of interestingness/usefulness/relevancy, which in my experience doesn't hold [...] Are you really driver by those numbers, or do you use them as a hint of how valuable the feed is?

Main idea is: articles of a rarely posting feed shouldn't be "suppressed" / get hidden by a feed posting articles with much higher frequency. It has nothing to do with using posting frequency as signal of usefulness or relevancy!

Let me explain, how I use Newsboat: I would like to have query feeds for latest unread articles per topic - like *Latest/Code* or *Latest/News*. I don't want to navigate to every single feed and look for latest articles - with a growing number of feeds, this would get more and more cumbersome.

Though, If you have all topic-related content in one query, it gets much harder to discover articles from rarely posting feeds, if activity of another feed(s) is high enough. Imagine Feed A has 500 articles per week, Feed B has 1 article per month, both in one query. Good luck finding articles of B :-).

> I encountered total_count, but this attribute seems to take into account only the total count of all feeds.

If you want to have further infos: [this medium article](https://medium.com/the-sandbox-planet-project/i-have-been-using-feedly-wrong-for-5-years-its-fixed-now-6b5fd7d3a4e3) has been one driver/motivation to organize categories by their frequency for me.

**Comment by f6ADt** [comment]
@Minoru concerning `total_count`: You are right! This has been a typo, I'll fix that in OP. I meant:

> `total_count`, but this attribute seems to take into account only the **total count of articles per feed**.

**Comment by Minoru** [comment]
> (later on other options possible)

No, "more options later" is exactly what I'm trying to *avoid* :) Newsboat has way too many knobs already, and I'd be a fool to set myself up for adding even more in the future. It's better to think this through at least a little bit.

Good point regarding dead feeds. Let's keep them out of this discussion then.

> I don't want to navigate to every single feed and look for latest articles - with a growing number of feeds, this would get more and more cumbersome.

You don't have to: navigate to any one of them, and start tapping `n` (which executes `next-unread` operation). When you're done with unread articles in the current feed, Newsboat will jump to the next that has more unread articles. Do read the manual, there's plenty of nifty keys to discover ;)

After reading the Medium article, I'm even more convinced that this is entirely about using frequency as a signal of usefulness. The author over-subscribed, and now tries different tricks to manage the load. Instead of giving each article an equal share of attention, they decided to give each *class* of feeds an equal share of attention. The classes are defined the posting frequency. In other words, they're saying that each article in "Super Fast" feed is *less* deserving of their attention than an article from a "Slow" feed.

Frankly, I'm not too excited about implementing all of the aforementioned additions that would be needed for this feature to be useful. I'll leave this in the "to consider" pile for now, and see if anyone else stumbles into this and describes more ways how it can be useful.

**Comment by f6ADt** [comment]
> After reading the Medium article, I'm even more convinced that this is entirely about using frequency as a signal of usefulness. 

I likely misunderstood your statement in the previous post.  By "posting frequency as a signal of interestingness/usefulness/relevancy" you meant that each article does not receive equal attention anymore? Yes, that is true. I would rather skim through a news feed with 500 articles per week, but would want to go through all my blog gems.

Shortcut `n` seems like a good idea, in addition to `A` to mark the whole news feed as read after having read some of its articles. I'll try that out, thanks!
